### PR TITLE
[Large Tensor] Fix cumsum op

### PR DESCRIPTION
**Pull request by connorgoggins**
## Description ##
The cumsum op was previously breaking on large tensor (dimension > 2^32) data. With the following input:
```
run_performance_test(nd.cumsum, inputs=[{"a": nd.random_normal(shape=(2**32 + 1, 1))}], run_backward=True, warmup=1, runs=1)
```
the following error was thrown:
```
Segmentation fault (core dumped)
```

To root cause this issue, I ran the previous command in a Python script with GDB, and found that the underlying problem was in the data type of several variables in the forward/backward structs in `np_cumsum-inl.h.h`. These variables used the `int` dtype when they should have been using `index_t` to properly handle long int indices. I switched these variables to `index_t` in the struct header and, after rebuilding, the previous input command displayed the correct output:
```
INFO:root:Begin Benchmark - cumsum
INFO:root:Complete Benchmark - cumsum
[{'cumsum': [{'inputs': {'a': '<NDArray 4294967297x1 @cpu(0)>'}, 'max_storage_mem_alloc_cpu/0': 33285996.0, 'avg_time_forward_cumsum': 4366.7148, 'avg_time_backward_cumsum': 12744.9971}]}]
```

To ensure completeness and to prevent future breaking changes, I also added a nightly test for the cumsum op with large tensor data in `tests/nightly/test_large_array.py`.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- M src/operator/numpy/np_cumsum-inl.h
- M tests/nightly/test_large_array.py

## Comments ##
Tested on r5dn.24xl-ubuntu 16.04 with
1. Individual op run
2. Full OpPerf run

## Results ##
[Single operator test - cumsum op (CPU)](https://gist.github.com/connorgoggins/7d1a5912512b1dbbacff350e3cd576ee)

[Full OpPerf test (CPU)](https://gist.github.com/connorgoggins/4538df906c3fbe36f989bb9067b8feb8)

@apeforest @access2rohit @ChaiBapchya

### CODE REVIEW

**Comment by connorgoggins** [comment]
@mxnet-label-bot add [pr-awaiting-review]

File: src/operator/numpy/np_cumsum-inl.h [hunk 1]
```diff
@@ -60,17 +60,17 @@ struct CumsumParam : public dmlc::Parameter<CumsumParam> {
...
                                   const IType *in,
-                                  const int middle,
-                                  const int trailing) {
-    int left = i / trailing, right = i % trailing;
-    int offset = left * middle * trailing + right;
+                                  const index_t middle,
+                                  const index_t trailing) {
+    index_t left = i / trailing, right = i % trailing;
+    index_t offset = left * middle * trailing + right;
     const IType *lane_in = in + offset;
```
Comment (access2rohit) [inline src/operator/numpy/np_cumsum-inl.h]: possible values for IType ? Since this can also be an iterator

File: src/operator/numpy/np_cumsum-inl.h [hunk 2]
```diff
@@ -125,17 +125,17 @@ void CumsumForward(const nnvm::NodeAttrs& attrs,
...
-                                  const int middle,
-                                  const int trailing) {
-    int left = i / trailing, right = i % trailing;
-    int offset = left * middle * trailing + right;
+                                  const index_t middle,
+                                  const index_t trailing) {
+    index_t left = i / trailing, right = i % trailing;
+    index_t offset = left * middle * trailing + right;
     const OType *lane_ograd = ograd + offset;
     IType *lane_igrad = igrad + offset;
```
Comment (access2rohit) [inline src/operator/numpy/np_cumsum-inl.h]: same as above

File: src/operator/numpy/np_cumsum-inl.h (hunk 1 above)
Comment (connorgoggins) [reply src/operator/numpy/np_cumsum-inl.h]: [Here](https://www.cct.lsu.edu/~rguidry/eclipse-doc36/org/eclipse/cdt/core/dom/ast/IType.html) it's defined as an interface for all C and C++ types.

**Comment by sandeep-krishnamurthy** [comment]
1. Are there any performance regression associated with this index type change? (Performance on Before and after the change)
2. Thanks for fix this issue, is there a way to know if there are any other broken operators with similar issue? (Not intended to fix with this PR, but, wanted to see the blast radius and possibly create a tracking issue while we are on it here)

**Comment by connorgoggins** [comment]
@sandeep-krishnamurthy thanks for your questions!

1. If you look at performance of the `cumsum` op [before](https://gist.github.com/connorgoggins/3fc3c7a40dae8b5cea7f82d9ca4b1a72) vs. [after](https://gist.github.com/connorgoggins/4538df906c3fbe36f989bb9067b8feb8) the changes in this PR, you can see that there is no regression on the time required for a forward pass.
2. My current assignment has been to systematically fix all ops with similar issues (i.e. throwing errors when handling large tensor data). As of now, I have created PRs with fixes for every problematic op, and once all my remaining op fix PRs (3 remaining PRs, including this one) are merged, all similar issues will be resolved.

File: src/operator/numpy/np_cumsum-inl.h [hunk 3]
```diff
@@ -98,11 +98,11 @@ void CumsumForwardImpl(const OpContext& ctx,
   }
 
   Stream<xpu> *s = ctx.get_stream<xpu>();
-  MSHADOW_TYPE_SWITCH_WITH_BOOL(in.type_flag_, IType, {
+  MSHADOW_TYPE_SWITCH_WITH_BOOL(in.type_flag_, index_t, {
```
Comment (apeforest) [inline src/operator/numpy/np_cumsum-inl.h]: I think you will not need this switch macro, right?

File: src/operator/numpy/np_cumsum-inl.h [hunk 4]
```diff
@@ -157,10 +157,10 @@ void CumsumBackwardImpl(const OpContext& ctx,
     }
   }
   Stream<xpu> *s = ctx.get_stream<xpu>();
-  MSHADOW_TYPE_SWITCH_WITH_BOOL(igrad.type_flag_, IType, {
+  MSHADOW_TYPE_SWITCH_WITH_BOOL(igrad.type_flag_, index_t, {
```
Comment (access2rohit) [inline src/operator/numpy/np_cumsum-inl.h]: Can you revert this change as well ? rest LGTM !

File: src/operator/numpy/np_cumsum-inl.h [hunk 5]
```diff
@@ -157,10 +157,10 @@ void CumsumBackwardImpl(const OpContext& ctx,
...
     MSHADOW_TYPE_SWITCH(ograd.type_flag_, OType, {
       Kernel<cumsum_backward, xpu>::Launch(
-        s, igrad.Size() / middle, igrad.dptr<IType>(),
+        s, igrad.Size() / middle, igrad.dptr<index_t>(),
```
Comment (access2rohit) [inline src/operator/numpy/np_cumsum-inl.h]: Ditto

File: src/operator/numpy/np_cumsum-inl.h (hunk 4 above)
Comment (apeforest) [inline src/operator/numpy/np_cumsum-inl.h]: You don't need a macro if the type is predefined.

File: src/operator/numpy/np_cumsum-inl.h (hunk 3 above)
Comment (connorgoggins) [reply src/operator/numpy/np_cumsum-inl.h]: Great point, reverted to IType.

File: src/operator/numpy/np_cumsum-inl.h (hunk 4 above)
Comment (connorgoggins) [reply src/operator/numpy/np_cumsum-inl.h]: Great point, reverted to IType.

File: src/operator/numpy/np_cumsum-inl.h (hunk 5 above)
Comment (connorgoggins) [reply src/operator/numpy/np_cumsum-inl.h]: Reverted, thanks!
